File: plugins/lofi/bitcrusher/bitcrusher.js
// Bit Crusher — quantises sample depth and downsamples by holding each
// sample for N hops. Both knobs together give the chiptune / Atari
// flavor; just the bit-depth knob is enough for soft lo-fi crunch.

function BitCrusherProcessor() {
  this.holdCounter = 0;
  this.held = 0;
}
BitCrusherProcessor.parameterDescriptors = [
  { name: 'bits', label: 'Bits',       defaultValue: 8, minValue: 1,  maxValue: 16 },
  { name: 'rate', label: 'Hold (smp)', defaultValue: 1, minValue: 1,  maxValue: 64 }
];
BitCrusherProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var bits = Math.max(1, Math.floor(parameters.bits[0]));
  var hold = Math.max(1, Math.floor(parameters.rate[0]));
  var steps = Math.pow(2, bits - 1);
  for (var i = 0; i < input.length; i++) {
    if (this.holdCounter <= 0) {
      // Quantise to `bits`-bit resolution.
      this.held = Math.round(input[i] * steps) / steps;
      this.holdCounter = hold;
    }
    output[i] = this.held;
    this.holdCounter--;
  }
  return true;
};

registerProcessor('bitcrusher', BitCrusherProcessor);

File: plugins/lofi/bitcrusher/Bitcrusher.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Bit Crusher (native port) — quantises sample depth + holds each sample
// for N hops. Chiptune crunch.
public final class Bitcrusher implements VocalMonitorNativePlugin {
    private int holdCounter = 0;
    private float held = 0f;
    private float bits = 8f, rate = 1f;

    @Override public void init(int sr) { holdCounter = 0; held = 0f; }
    @Override public String[] parameterNames() { return new String[] { "bits", "rate" }; }
    @Override public float parameterMin(String n) { return 1f; }
    @Override public float parameterMax(String n) { return "bits".equals(n) ? 16f : 64f; }
    @Override public float parameterDefault(String n) { return "bits".equals(n) ? 8f : 1f; }
    @Override public String parameterLabel(String n) { return "bits".equals(n) ? "Bits" : "Hold (smp)"; }
    @Override public void setParameter(String n, float v) {
        if ("bits".equals(n)) bits = v;
        else if ("rate".equals(n)) rate = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final int bitsInt = Math.max(1, (int) Math.floor(bits));
        final int hold = Math.max(1, (int) Math.floor(rate));
        final float steps = (float) Math.pow(2.0, bitsInt - 1);
        final float invSteps = 1f / steps;
        int hc = holdCounter;
        float h = held;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            if (hc <= 0) {
                h = Math.round(input[i] * steps) * invSteps;
                hc = hold;
            }
            output[i] = h;
            hc--;
        }
        holdCounter = hc;
        held = h;
    }
}
